pkg/server/image_pull.go

```diff
@@ -17,14 +17,315 @@ limitations under the License.
 package server
 
 import (
-	"errors"
+	gocontext "context"
+	"encoding/json"
+	"fmt"
+	"github.com/golang/glog"
+	"path/filepath"
+	"sync"
+	"time"
 
+	"github.com/containerd/containerd/content"
+	containerd_images "github.com/containerd/containerd/images"
+	"github.com/containerd/containerd/remotes"
+	"github.com/containerd/containerd/remotes/docker"
+	"github.com/kubernetes-incubator/cri-containerd/pkg/metadata"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
 	"golang.org/x/net/context"
-
+	"golang.org/x/sync/errgroup"
 	"k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime"
 )
 
 // PullImage pulls an image with authentication config.
+// TODO add authentication
+// TODO harden api
 func (c *criContainerdService) PullImage(ctx context.Context, r *runtime.PullImageRequest) (*runtime.PullImageResponse, error) {
-	return nil, errors.New("not implemented")
+	var (
```

Comment:
Will add a TODO to figure out at what layer we should be blocking on duplicate requests.